x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/support/SamlObjectSigner.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.idp.saml.support;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.xpack.idp.saml.idp.SamlIdentityProvider;
+import org.opensaml.xmlsec.signature.SignableXMLObject;
+import org.opensaml.xmlsec.signature.Signature;
+import org.opensaml.xmlsec.signature.support.SignatureConstants;
+import org.opensaml.xmlsec.signature.support.SignatureException;
+import org.opensaml.xmlsec.signature.support.Signer;
+import org.w3c.dom.Element;
+
+/**
+ * Signs OpenSAML {@link SignableXMLObject} instances using {@link SamlIdentityProvider#getSigningCredential()}.
+ */
+public class SamlObjectSigner {
+
+    private final SamlFactory samlFactory;
+    private final SamlIdentityProvider idp;
+
+    public SamlObjectSigner(SamlFactory samlFactory, SamlIdentityProvider idp) {
+        this.samlFactory = samlFactory;
+        this.idp = idp;
+    }
+
+    public Element sign(SignableXMLObject object) {
+        final Signature signature = samlFactory.object(Signature.class, Signature.DEFAULT_ELEMENT_NAME);
+        signature.setSigningCredential(idp.getSigningCredential());
+        signature.setSignatureAlgorithm(SignatureConstants.ALGO_ID_SIGNATURE_RSA_SHA256);
+        signature.setCanonicalizationAlgorithm(SignatureConstants.ALGO_ID_C14N_EXCL_OMIT_COMMENTS);
+        object.setSignature(signature);
+        Element element = samlFactory.toDomElement(object);
```

Comment:
This is true, but the `Signature` on the `SignableXMLObject` will still be null.

So you can keep working with the OpenSAML layer objects (because they have a copy of the DOM element with the signature), but it's confusing because the OpenSAML model is no longer an accurate reflection of the DOM state (one has signature data, the other doesn't).

I prefer to reflect that by returning `Element` because that is the only model that clearly reflects the newly attached signature.

Reply:
> This is true, but the Signature on the SignableXMLObject will still be null.

But don't we 
```
object.setSignature(signature);
```

right above ? I'm fine to merge as is, I don't even know if we get so much by not returning `Element` here, just for the sake of discussion.

Reply:
You're right. Thanks for keeping me honest.